osmoutils/accum/accum_test.go

```diff
@@ -318,3 +321,439 @@ func (suite *AccumTestSuite) TestClaimRewards() {
+			startingNumShares: initialValueOne,
+			startingUnclaimedRewards: sdk.NewDecCoins(),
+			newShares: sdk.NewDecWithPrec(983429874321, 5),
+			accumInit: sdk.NewDecCoins(),
+			accumDelta: sdk.NewDecCoins(initialCoinDenomOne, initialCoinDenomTwo),
+			expPass: true,
+		},
+
+		// error catching
+		"account does not exist": {
+			addrDNE: true,
+			expPass: false,
+
+			startingNumShares: sdk.ZeroDec(),
+			startingUnclaimedRewards: sdk.NewDecCoins(),
+			newShares: sdk.OneDec(),
+			accumInit: sdk.NewDecCoins(),
+			accumDelta: sdk.NewDecCoins(),
+		},
+		"attempt to add zero shares": {
+			startingNumShares: initialValueOne,
+			startingUnclaimedRewards: sdk.NewDecCoins(),
+			newShares: sdk.ZeroDec(),
+			accumInit: sdk.NewDecCoins(),
+			accumDelta: sdk.NewDecCoins(),
+			expPass: false,
+		},
+		"attempt to add negative shares": {
+			startingNumShares: initialValueOne,
+			startingUnclaimedRewards: sdk.NewDecCoins(),
+			newShares: sdk.OneDec().Neg(),
+			accumInit: sdk.NewDecCoins(),
+			accumDelta: sdk.NewDecCoins(),
+			expPass: false,
+		},
+	}
+
+	for name, tc := range tests {
+		suite.Run(name, func() {
+			// We reset the store for each test
+			suite.SetupTest()
+			addr := osmoutils.CreateRandomAccounts(1)[0]
+
+			// Create a new accumulator with initial value specified by test case
+			curAccum := accumPackage.CreateRawAccumObject(suite.store, testNameOne, tc.accumInit)
+
+			// Create new position in store (raw to minimize dependencies)
+			if !tc.addrDNE {
+				accumPackage.CreateRawPosition(curAccum, addr, tc.startingNumShares, tc.startingUnclaimedRewards, emptyPositionOptions)
+			}
+
+			// Update accumulator with accumDelta (increasing position's rewards by a proportional amount)
+			curAccum = accumPackage.CreateRawAccumObject(suite.store, testNameOne, tc.accumInit.Add(tc.accumDelta...))
+
+			// Add newShares to position
+			err := curAccum.AddToPosition(addr, tc.newShares)
+
+			if tc.expPass {
+				suite.Require().NoError(err)
+
+				// Get updated position for comparison
+				newPosition, err := accumPackage.GetPosition(curAccum, addr)
+				suite.Require().NoError(err)
+
+				// Ensure position's accumulator value is moved up to init + delta
+				suite.Require().Equal(tc.accumInit.Add(tc.accumDelta...), newPosition.InitAccumValue)
+
+				// Ensure accrued rewards are moved into UnclaimedRewards (both when it starts empty and not)
+				suite.Require().Equal(tc.startingUnclaimedRewards.Add(tc.accumDelta.MulDec(tc.startingNumShares)...), newPosition.UnclaimedRewards)
+
+				// Ensure address's position properly reflects new number of shares
+				suite.Require().Equal(tc.startingNumShares.Add(tc.newShares), newPosition.NumShares)
+
+				// Ensure a new position isn't created in memory (only old one is overwritten)
+				allAccumPositions, err := curAccum.GetAllPositions()
+				suite.Require().NoError(err)
+				suite.Require().True(len(allAccumPositions) == 1)
+			} else {
+				suite.Require().Error(err)
```

Comment:
Should we add a validation that if an error occurs, no new shares are actually added to any position?

Reply:
Good call, will add this